src/display/display_utils.js

```diff
@@ -1124,6 +1124,36 @@ function setLayerDimensions(
   }
 }
 
+/**
+ * Scale factors for the canvas, necessary with HiDPI displays.
+ */
+class OutputScale {
+  constructor() {
+    const pixelRatio = window.devicePixelRatio || 1;
+
+    /**
+     * @type {number} Horizontal scale.
+     */
+    this.sx = pixelRatio;
+
+    /**
+     * @type {number} Vertical scale.
+     */
+    this.sy = pixelRatio;
+  }
+
+  /**
+   * @type {boolean} Returns `true` when scaling is required, `false` otherwise.
+   */
+  get scaled() {
+    return this.sx !== 1 || this.sy !== 1;
+  }
+
+  get homothetic() {
```

Comment:
It might just be being uneducated, but I had to lookup this word :-)
Could we perhaps pick a different, slightly "easier", name here; i.e. would the following convey approximately the same thing?

```suggestion
  get symmetric() {
```